dvc/fs/dvc.py

```diff
@@ -333,21 +333,20 @@ def ls(  # pylint: disable=arguments-differ # noqa: C901
                     fs, fs_path, detail=True, ignore_subrepos=ignore_subrepos
                 ):
                     fs_infos[fs.path.name(info["name"])] = info
+                fs_exists = True
             except (FileNotFoundError, NotADirectoryError):
                 pass
 
-            fs_exists = bool(fs_infos) or fs.exists(fs_path)
+        if not (dvc_exists or fs_exists):
+            # broken symlink or TreeError
+            raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), path)
```

Comment:
Want to raise an error if no path found in either filesystem. Moved this up since it doesn't actually depend on `names` (which was used in the deleted logic below).